File: src/embedder.ts
/**
 * Embedding pipeline — EmbeddingGemma-300M via node-llama-cpp.
 *
 * Auto-downloads the GGUF model from HuggingFace on first use.
 * Runs on Apple Silicon Metal for GPU acceleration.
 */

import {
  getLlama,
  type Llama,
  type LlamaModel,
  type LlamaEmbeddingContext,
} from "node-llama-cpp";

const MODEL_URI =
  "hf:ggml-org/embeddinggemma-300M-GGUF/embeddinggemma-300M-Q8_0.gguf";

const VECTOR_DIM = 768;

let _llama: Llama | null = null;
let _model: LlamaModel | null = null;
let _ctx: LlamaEmbeddingContext | null = null;
let _initPromise: Promise<void> | null = null;

/**
 * Initialize the embedding model (lazy, singleton).
 * Downloads ~314MB on first run.
 */
async function ensureModel(): Promise<LlamaEmbeddingContext> {
  if (_ctx) return _ctx;

  if (!_initPromise) {
    _initPromise = (async () => {
      _llama = await getLlama();
      _model = await _llama.loadModel({ modelPath: MODEL_URI });
      _ctx = await _model.createEmbeddingContext({
        contextSize: 2048,
      });
    })();
  }

  await _initPromise;
  return _ctx!;
}

/**
 * Embed a single text string. Returns a Float32Array of dimension 768.
 */
export async function embed(text: string): Promise<Float32Array> {
  const ctx = await ensureModel();
  const result = await ctx.getEmbeddingFor(text);
  return new Float32Array(result.vector);
}

/**
 * Embed multiple texts in batch.
 */
export async function embedBatch(texts: string[]): Promise<Float32Array[]> {
  const results: Float32Array[] = [];
  for (const text of texts) {
    results.push(await embed(text));
  }
  return results;
}

/**
 * Split a document into fragments for embedding.
 * Embeds the full document plus individual sections.
 */
export function fragmentize(title: string, content: string): Array<{ seq: number; text: string }> {
  const fragments: Array<{ seq: number; text: string }> = [];

  // seq=0: full document (title + content, truncated)
  const fullText = `title: ${title} | text: ${content}`.slice(0, 2000);
  fragments.push({ seq: 0, text: fullText });

  // Split by markdown headers or double newlines
  const sections = content.split(/(?=^#{1,3}\s)/m).filter((s) => s.trim().length > 50);

  for (let i = 0; i < Math.min(sections.length, 5); i++) {
    fragments.push({
      seq: i + 1,
      text: `title: ${title} | section: ${sections[i]!.trim().slice(0, 1000)}`,
    });
  }

  return fragments;
}

/**
 * Embed and store all fragments for a document.
 */
export async function embedDocument(
  store: { saveEmbedding: (hash: string, seq: number, emb: Float32Array) => void },
  contentHash: string,
  title: string,
  content: string,
): Promise<number> {
  const fragments = fragmentize(title, content);
  let count = 0;

  for (const frag of fragments) {
    const emb = await embed(frag.text);
    store.saveEmbedding(contentHash, frag.seq, emb);
    count++;
  }

  return count;
}

export { VECTOR_DIM };

File: src/mnemon/embedder.py
"""Embedding pipeline — FastEmbed with bge-small-en-v1.5 (ONNX).

384-dimensional embeddings via ONNX Runtime. ~13MB model, auto-downloaded.
No PyTorch dependency needed.
"""

from __future__ import annotations

import re
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import numpy as np
    from .store import Store

VECTOR_DIM = 384
_MODEL_NAME = "BAAI/bge-small-en-v1.5"

_model = None


def _get_model():
    """Lazy-load the embedding model (singleton)."""
    global _model
    if _model is None:
        from fastembed import TextEmbedding
        _model = TextEmbedding(model_name=_MODEL_NAME)
    return _model


def embed(text: str) -> "np.ndarray":
    """Embed a single text string. Returns ndarray of shape (384,)."""
    import numpy as np
    model = _get_model()
    result = list(model.embed([text]))
    return np.asarray(result[0], dtype=np.float32)


def embed_batch(texts: list[str]) -> list["np.ndarray"]:
    """Embed multiple texts."""
    import numpy as np
    model = _get_model()
    results = list(model.embed(texts))
    return [np.asarray(r, dtype=np.float32) for r in results]


def fragmentize(title: str, content: str) -> list[dict]:
    """Split a document into fragments for embedding.

    Returns list of {seq, text} dicts:
      seq=0: full document (title + content, truncated to 2000 chars)
      seq=1-5: individual sections split by markdown headers or double newlines
    """
    fragments = []

    # seq=0: full document
    full_text = f"title: {title} | text: {content}"[:2000]
    fragments.append({"seq": 0, "text": full_text})

    # Split by markdown headers or double newlines
    sections = re.split(r"(?=^#{1,3}\s)", content, flags=re.MULTILINE)
    sections = [s.strip() for s in sections if len(s.strip()) > 50]

    for i, section in enumerate(sections[:5]):
        fragments.append({
            "seq": i + 1,
            "text": f"title: {title} | section: {section[:1000]}",
        })

    return fragments


def embed_document(store: "Store", content_hash: str, title: str, content: str) -> int:
    """Embed and store all fragments for a document. Returns fragment count."""
    fragments = fragmentize(title, content)
    count = 0

    for frag in fragments:
        emb = embed(frag["text"])
        store.save_embedding(content_hash, frag["seq"], emb)
        count += 1

    store.flush_vectors()
    return count
